### ISSUE TITLE
Question: Can I enable numlock and can I switch the desktop mode?

### ISSUE THREAD

**Issue by Svenum** [nix-community/plasma-manager#117]
1. I would like to change the default behave from the numlock with this flake, so that it gets reactivated on every login, is this possible?
   Setting in GUI: Open Plasma Settings -> Keyboard -> Keyboard -> Hardware -> NumLock on Plasma Startup -> Turn on

2. I would like to be able to change the desktop mode from "Folder View" to "Desktop" with this module, is this possible?
   Setting in GUI: Right Click on Desktop -> Configure Desktop and Wallpaper -> Layout -> Select "Desktop" in Dropdown Menu

**Comment by magnouvean** [comment]
Does setting `programs.plasma.configFile.kcminputrc.Keyboard.NumLock.value = 0;` fix this maybe?

**Comment by Svenum** [comment]
This seems to work, thanks.
Do you have an solution to the second question?

**Comment by magnouvean** [comment]
There it is harder. The relevant key seems to be the `plugin` key for each of the containments in `plasma-org.kde.plasma.desktop-appletsrc`, where the value is set to `org.kde.plasma.folder` for folderview and `org.kde.plasma.desktopcontainment` for desktop view. This probably will need a higher level interface and some plasma desktop scripting, so it isn't supported a.t.m..

**Comment by Svenum** [comment]
Could that be done by the panels? This is the type of scripting you mean? Right?

**Comment by magnouvean** [comment]
Yeah the same type of desktop scripting. Panels is one of the uses for it, but there are some more uses as well. I think desktop scripting is our best shot for getting this functionality.

**Comment by magnouvean** [comment]
Just tried the desktop mode thing with plasma script, and while it succeeds in writing the correct config to the file it doesn't get applied immediately, and hence probably would need a log out/in again, which doesn't work when we are running the script in the autostart. I'm afraid this doesn't have any elegant solution for now.

**Comment by rachelambda** [comment]
Hi, is there is still no fix for the desktop view? I'd love even something hacky, its the one thing missing from my original plasma config currently. If not, is there some reference for how plasma config works, so one could try to contribute?

**Comment by magnouvean** [comment]
The appropriate file this is configured in is `~/.config/plasma-org.kde.plasma.desktop-appletsrc`. The only good interface to modify that we have found is through [plasma desktop scripts](https://develop.kde.org/docs/plasma/scripting/), but for some reason after running the script required to configure this, the configuration isn't applied until plasma has been restarted (or some component of plasma at the very least). I can test again on plasma 6.1 if it has changed, though I kinda doubt it tbh. After #222 maybe we could implement a hacky solution though by restarting `plasma-desktop` after running the script. I can test this soon. If this doesn't work either I'm afraid we are out of luck until some changes are made on kde's side.

**Comment by magnouvean** [comment]
OK so I did try this, but unfortunately to no luck. What I have found out is that it appears that the desktop containment doesn't just have it's `plugin` key altered, but it appears that kde generates a completely new desktop-containment.

I don't really have that much time available for `plasma-manager` in the near future for personal reasons, so I'm not sure I'll figure this out. It's possible that a solution does exist though. If you (or anyone for that matter) want to experiment with this I would suggest trying to find some plasma-script code which changes the layout of a desktop to desktop-mode (you can run plasma-scripts in `plasma-interactiveconsole`). It's fine if the script only is successful after running `systemctl --user restart plasma-plasmashell` (or any other systemd service for that matter). What I have tried to no success is simply:
```javascript
for (var desktop of desktops()) {
  desktop.writeConfig("plugin", "org.kde.desktopcontainment")
}
```

**Comment by rachelambda** [comment]
thanks for trying! I will play around with this a bit in the days to come

**Comment by frantisekhanzlikbl** [comment]
I was hoping to come up with a better solution before posting here, but it's increasingly becoming apparent that I'll never ever find the time for that. 😅 Soooo, for those that don't mind

- A system closure larger by ~50 MB, and
- Rebuilds of everything that depends on `kdePackages.plasma-desktop` (a few (tens?) of packages, amounting to about 5 minutes of extra build time on my machine),

here is a piece of NixOS module code that works well enough:

```nix
nixpkgs.overlays = [
	(final: prev: {
		kdePackages = prev.kdePackages.overrideScope (kpFinal: kpPrev: {
			plasma-desktop = final.applyPatches {
				src = kpPrev.plasma-desktop;
				patches = [ (builtins.toFile "plasma-desktop-shell-defaults.patch" ''
					diff --git a/share/plasma/shells/org.kde.plasma.desktop/contents/defaults b/share/plasma/shells/org.kde.plasma.desktop/contents/defaults
					index 63c6b2b..b9ebdc7 100644
					--- a/share/plasma/shells/org.kde.plasma.desktop/contents/defaults
					+++ b/share/plasma/shells/org.kde.plasma.desktop/contents/defaults
					@@ -1,5 +1,5 @@
					 [Desktop]
					-Containment=org.kde.plasma.folder
					+Containment=org.kde.desktopcontainment
					 ToolBox=
					 RuntimePlatform=Desktop
				'') ];
			};
		});
	})
];
```
(hopefully the patch code will get properly stripped—my real setup is slightly different. if not, move the patch to a file)

I'm not sure whether just updating the defaults makes KDE change existing configurations, so performing the change manually (either in the GUI, manually in `~/.config/plasma-org.kde.plasma.desktop-appletsrc`, or just by nuking that file) on existing systems might be necessary as well. Hopefully at least this is useful to somebody! ^-^
